grafana/grafana-operator
# add github datasource

## Description

Allows Operator to support GitHub datasources by creating a GrafanaDataSource.
## Relevant issues/tickets

Fixes https://github.com/grafana-operator/grafana-operator/issues/570
## Type of change




## Checklist


## Verification steps

Verifying locally by CRC
1. Create the following resources to be applied
  * Secret to install GitHub plugin
```
apiVersion: v1
kind: Secret
metadata:
  name: grafana-env
stringData:
  GF_INSTALL_PLUGINS: grafana-github-datasource 1.0.6
type: Opaque
```
   * Grafana CR for GitHub plugin
```
apiVersion: integreatly.org/v1alpha1
kind: Grafana
metadata:
  name: example-grafana
spec:
  client:
    preferService: false
  deployment:
    envFrom:
      - secretRef:
          name: grafana-env
  ingress:
    enabled: True
  config:
    log:
      mode: "console"
      level: "warn"
    auth:
      disable_login_form: False
      disable_signout_menu: True
    auth.anonymous:
      enabled: True
  dashboardLabelSelector:
    - matchExpressions:
        - { key: app, operator: In, values: [ grafana ] }
```
   * GitHub dashboard (using sample dashboard found here)
```
apiVersion: integreatly.org/v1alpha1
kind: GrafanaDashboard
metadata:
  name: github-dashboard
  labels:
    app: grafana
spec:
  json:
    ""
  url: https://raw.githubusercontent.com/grafana/github-datasource/master/src/dashboards/dashboard.json
  datasources:
    - inputName: "DS_GITHUB"
      datasourceName: GitHub
  plugins:
    - name: "grafana-github-datasource"
      version: "1.0.6"
```
   * GitHub Datasource (this example can also be found here)
```
apiVersion: integreatly.org/v1alpha1
kind: GrafanaDataSource
metadata:
  name: github-grafanadatasource
spec:
  name: github-grafanadatasource.yaml
  datasources:
    - name: GitHub
      type: grafana-github-datasource
      access: proxy
      jsonData:
        githubUrl: ""
      secureJsonData:
        accessToken: <github-token>

```

>Note: The GitHub token provided must have the following permissions:
>For all repositories:
>* public_repo
>* repo:status
>* repo_deployment
>* read:packages
>* user:read
>* user:email
>
>An extra setting is required for private repositories
>* repo (Full control of private repositories)

2. Run the Operator locally using this branch.
3. Apply the resources to a namespace. 
4. WIth CRC, a Grafana route should be created. Open the route location to open Grafana UI.
5. Navigate to the dashboard titled `GitHub Default` (Found under Dashboards/Manage in the side bar)
6. Dashboard should display data from `https://github.com/grafana/grafana` by default. Different repos can be chosen by editting the `Organization` and `Repository` fields in the UI.
